### PR TITLE
Break cycles with Dune_load and Inline_tests

### PR DESCRIPTION
**Pull request by cwong-ocaml**
Refactoring only. Loading subsystems should be less cycle-prone now.

Signed-off-by: Cameron Wong <cwong@janestreet.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
Could you describe the cycle in a little more detail?

File: src/dune_rules/dune_loaded_file.mli [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+open! Dune_engine
+open! Stdune
+
+(** A fully evaluated dune file *)
+type t =
+  { dir : Path.Source.t
+  ; project : Dune_project.t
+  ; stanzas : Dune_file.Stanzas.t
+  }
```
Comment (rgrinberg) [inline src/dune_rules/dune_loaded_file.mli:11]: You could also move this to `Dune_file`, so the the type can be `Dune_file.t` I would then prefer to start splitting away other stanza components from `Dune_file`.

**Comment by cwong-ocaml** [comment]
The cycle isn't present on main, but as part of our transition within Jane Street, we are reading jbuild files and converting them to a `Dune_file.Stanzas.t` on the fly. This ended up inducing a cycle dune_load -> jbuild_converter -> dune_file -> dune_load (and another, more complex one with `inline_test.ml` whose details elude me at present).

**Comment by rgrinberg** [comment]
I see. That makes sense. The name I suggested makes a little more sense IMO however.

File: src/dune_rules/dune_loaded_file.mli (hunk 1 above)
Comment (cwong-ocaml) [reply src/dune_rules/dune_loaded_file.mli:11]: What would we name the existing `Dune_file` module in dune_rules? I'm not opposed to this change in principle, but I think I'm misunderstanding what you're asking for

Comment (aalekseyev) [reply src/dune_rules/dune_loaded_file.mli:11]: @cwong-ocaml, I think Rudi proposes to move the record definition into the existing module, which seems reasonable to me.
Note that Dune_file.t currently does not exist, so it looks like destiny that the name was kept available. :-)

Comment (rgrinberg) [reply src/dune_rules/dune_loaded_file.mli:11]: Yep, this is what I meant. Thanks for clarifying.

Comment (cwong-ocaml) [reply src/dune_rules/dune_loaded_file.mli]: Done! That seems like a much better idea.

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**